text/0000-pub-restricted.md

```diff
@@ -0,0 +1,924 @@
+- Feature Name: pub_restricted
+- Start Date: 2015-12-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Expand the current `pub`/non-`pub` categorization of items with the
+ability to say "make this item visible *solely* to a (named) module
+tree."
+
+The current `crate` is one such tree, and would be expressed via:
+`pub(crate) item`. Other trees can be denoted via a path employed in a
+`use` statement, e.g. `pub(a::b) item`, or `pub(super) item`.
+
+# Motivation
+[motivation]: #motivation
+
+Right now, if you have a definition for an item `X` that you want to
+use in many places in a module tree, you can either 
+(1.) define `X` at the root of the tree as a non-`pub` item, or
+(2.) you can define `X` as a `pub` item in some submodule
+(and import into the root of the module tree via `use`).
+
+But: Sometimes neither of these options is really what you want.
+
+There are scenarios where developers would like an item to be visible
+to a particular module subtree (or a whole crate in its entirety), but
+it is not possible to move the item's (non-pub) definition to the root
+of that subtree (which would be the usual way to expose an item to a
+subtree without making it pub).
+
+If the definition of `X` itself needs access to other private items
+within a submodule of the tree, then `X` *cannot* be put at the root
+of the module tree. Illustration:
+
+```rust
+// Intent: `a` exports `I` and `foo`, but nothing else.
+pub mod a {
+    pub const I: i32 = 3;
+
+    // `semisecret` will be used "many" places within `a`, but
+    // is not meant to be exposed outside of `a`.
+    fn semisecret(x: i32) -> i32  { use self::b::c::J; x + J }
+
+    pub fn foo(y: i32) -> i32 { semisecret(I) + y }
+    pub fn bar(z: i32) -> i32 { semisecret(I) * z }
```

Comment:
i added `fn bar` to the example later (just so "many places" that `semisecret` would be called did actually mean "more than one place"), but forgot to update the comment above the `mod a`.